stdlib/src/memory/box.mojo

```diff
@@ -0,0 +1,128 @@
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions:
+# https://llvm.org/LICENSE.txt
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+from memory import UnsafePointer, stack_allocation, memcpy
+
+
+struct Box[T: AnyType]:
+    """A safe, owning, smart pointer.
+
+    This smart pointer is designed for cases where there is clear ownership
+    of the underlying data, and restricts access to it through the lifetime
+    system such that no more than one mutable alias for the underlying data
+    may exist.
+
+    Parameters:
+        T: The type to be stored in the Box[].
+    """
+
+    var _inner: UnsafePointer[T, AddressSpace.GENERIC]
+
+    fn __init__[T: Movable](inout self: Box[T], owned value: T):
+        """Construct a new Box[] by moving the passed value into a new backing allocation.
+
+        Parameters:
+            T: The type of the data to store. It is restricted to `Movable` here to allow efficient move construction.
+
+        Args:
+            value: The value to move into the Box[].
+        """
+        self._inner = UnsafePointer[T].alloc(1)
+        self._inner.init_pointee_move(value^)
+
+    fn __init__[T: ExplicitlyCopyable](inout self: Box[T], *, copy_value: T):
+        """Construct a new Box[] by explicitly copying the passed value into a new backing allocation.
+
+        Parameters:
+            T: The type of the data to store.
+
+        Args:
+            copy_value: The value to explicitly copy into the Box[].
+        """
+        self._inner = UnsafePointer[T].alloc(1)
+        self._inner.init_pointee_explicit_copy(copy_value)
+
+    # TODO: disambiguation and other niceties
+    #    fn __init__[
+    #        T: Copyable
+    #    ](inout self: Box[T], value: T):
+    #        self._inner = UnsafePointer[T].alloc(1)
+    #        self._inner.init_pointee_copy(value)
+
+    fn __init__[
+        T: ExplicitlyCopyable
+    ](inout self: Box[T], *, copy_box: Box[T],):
+        """Construct a new Box[] by explicitly copying the value from another Box[].
+
+        Parameters:
+            T: The type of the data to store.
+
+        Args:
+            copy_box: The Box[] to copy.
+        """
+        self.__init__(copy_value=copy_box[])
+
+    fn __moveinit__(inout self, owned existing: Self):
+        """Move this Box[].
+
+        Args:
+            existing: The value to move.
+        """
+        self._inner = existing._inner
+        existing._inner = UnsafePointer[T]()
+
+    fn __getitem__(
+        ref [_, AddressSpace.GENERIC._value.value]self
+    ) -> ref [__lifetime_of(self._inner), AddressSpace.GENERIC._value.value] T:
```

Comment:
I spoke with @lattner today about https://github.com/modularml/mojo/pull/3524#discussion_r1772307879.  The TL;DR is that `__lifetime_of(self._inner)` won't work as you'd expect (modeling the lifetime of the inner data) today.  Can you please change it to `__lifetime_of(self)`?  Chris has some ideas around this topic of indirect lifetimes that will get implemented at some point, but for now, just use `__lifetime_of(self)`.

Reply:
Oh, ok! I feel like I changed it at some point to deal with the `address_space` thing not propagating correctly. Lemme fix that up.